diff --git a/parser.c b/parser.c
index 372c2ba9..b505c16d 100644
--- a/parser.c
+++ b/parser.c
@@ -7586,7 +7586,7 @@ xmlHandleUndeclaredEntity(xmlParserCtxtPtr ctxt, const xmlChar *name) {
 
 static xmlEntityPtr
 xmlLookupGeneralEntity(xmlParserCtxtPtr ctxt, const xmlChar *name, int inAttr) {
-    xmlEntityPtr ent;
+    xmlEntityPtr ent = NULL;
 
     /*
      * Predefined entities override any extra definition
@@ -9331,7 +9331,7 @@ next_attr:
 	if (defaults != NULL) {
 	    for (i = 0; i < defaults->nbAttrs; i++) {
                 xmlDefAttr *attr = &defaults->attrs[i];
-                const xmlChar *nsuri;
+                const xmlChar *nsuri = NULL;
                 unsigned hashValue, uriHashValue;
                 int res;
 
diff --git a/xmlIO.c b/xmlIO.c
index 640d909c..02506207 100644
--- a/xmlIO.c
+++ b/xmlIO.c
@@ -2903,11 +2903,13 @@ xmlParserGetDirectory(const char *filename) {
         if (cur == dir) dir[1] = 0;
 	else *cur = 0;
 	ret = xmlMemStrdup(dir);
+#if !defined(WINAPI_FAMILY_PARTITION) || !WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_APP) || WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_DESKTOP)
     } else {
         if (getcwd(dir, 1024) != NULL) {
 	    dir[1023] = 0;
 	    ret = xmlMemStrdup(dir);
 	}
+#endif
     }
     return(ret);
 #undef IS_XMLPGD_SEP
diff --git a/xmlmodule.c b/xmlmodule.c
index dd2fd7a5..a80bac1f 100644
--- a/xmlmodule.c
+++ b/xmlmodule.c
@@ -267,7 +267,15 @@ xmlModulePlatformSymbol(void *handle, const char *name, void **symbol)
 static void *
 xmlModulePlatformOpen(const char *name)
 {
+#if defined(WINAPI_FAMILY_PARTITION) && WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_APP) && !WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_DESKTOP)
+    WCHAR wlpName[MAX_PATH];
+    if (MultiByteToWideChar(CP_UTF8, 0, name, -1, wlpName, MAX_PATH) == 0) {
+        return NULL;
+    }
+    return LoadPackagedLibrary(wlpName, 0);
+#else
     return LoadLibraryA(name);
+#endif
 }
 
 /*
diff --git a/xpath.c b/xpath.c
index 0ddde8d4..807a748d 100644
--- a/xpath.c
+++ b/xpath.c
@@ -8481,7 +8481,7 @@ void
 xmlXPathLangFunction(xmlXPathParserContextPtr ctxt, int nargs) {
     xmlXPathObjectPtr val;
     xmlNodePtr cur;
-    xmlChar *theLang;
+    xmlChar *theLang = NULL;
     const xmlChar *lang;
     int ret = 0;
     int i;
