diff --git a/globals.c b/globals.c
index 4726505b..8523b686 100644
--- a/globals.c
+++ b/globals.c
@@ -77,7 +77,9 @@ static xmlMutex xmlThrDefMutex;
 
 #ifdef HAVE_WIN32_THREADS
   #if defined(LIBXML_STATIC) && !defined(LIBXML_STATIC_FOR_DLL)
-    #define USE_WAIT_DTOR
+    #if !defined(WINAPI_FAMILY_PARTITION) || (!WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_APP) && WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_DESKTOP))
+      #define USE_WAIT_DTOR
+    #endif
   #else
     #define USE_DLL_MAIN
   #endif
diff --git a/xmlmodule.c b/xmlmodule.c
index 407227c1..1379fe14 100644
--- a/xmlmodule.c
+++ b/xmlmodule.c
@@ -263,7 +263,15 @@ xmlModulePlatformSymbol(void *handle, const char *name, void **symbol)
 static void *
 xmlModulePlatformOpen(const char *name)
 {
+#if defined(WINAPI_FAMILY_PARTITION) && WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_APP) && !WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_DESKTOP)
+    WCHAR wlpName[MAX_PATH];
+    if (MultiByteToWideChar(CP_UTF8, 0, name, -1, wlpName, MAX_PATH) == 0) {
+        return NULL;
+    }
+    return LoadPackagedLibrary(wlpName, 0);
+#else
     return LoadLibraryA(name);
+#endif
 }
 
 /*
diff --git a/xpath.c b/xpath.c
index 7be6cfdc..75b62a87 100644
--- a/xpath.c
+++ b/xpath.c
@@ -8203,7 +8203,7 @@ void
 xmlXPathLangFunction(xmlXPathParserContextPtr ctxt, int nargs) {
     xmlXPathObjectPtr val;
     xmlNodePtr cur;
-    xmlChar *theLang;
+    xmlChar *theLang = NULL;
     const xmlChar *lang;
     int ret = 0;
     int i;
