diff --git a/parser.c b/parser.c
index 44467355..092e8390 100644
--- a/parser.c
+++ b/parser.c
@@ -7612,7 +7612,7 @@ xmlHandleUndeclaredEntity(xmlParserCtxtPtr ctxt, const xmlChar *name) {
 
 static xmlEntityPtr
 xmlLookupGeneralEntity(xmlParserCtxtPtr ctxt, const xmlChar *name, int inAttr) {
-    xmlEntityPtr ent;
+    xmlEntityPtr ent = NULL;
 
     /*
      * Predefined entities override any extra definition
@@ -9390,7 +9390,7 @@ next_attr:
 	if (defaults != NULL) {
 	    for (i = 0; i < defaults->nbAttrs; i++) {
                 xmlDefAttr *attr = &defaults->attrs[i];
-                const xmlChar *nsuri;
+                const xmlChar *nsuri = NULL;
                 unsigned hashValue, uriHashValue;
                 int res;
 
diff --git a/xmlmodule.c b/xmlmodule.c
index dd2fd7a5..a80bac1f 100644
--- a/xmlmodule.c
+++ b/xmlmodule.c
@@ -267,7 +267,15 @@ xmlModulePlatformSymbol(void *handle, const char *name, void **symbol)
 static void *
 xmlModulePlatformOpen(const char *name)
 {
+#if defined(WINAPI_FAMILY_PARTITION) && WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_APP) && !WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_DESKTOP)
+    WCHAR wlpName[MAX_PATH];
+    if (MultiByteToWideChar(CP_UTF8, 0, name, -1, wlpName, MAX_PATH) == 0) {
+        return NULL;
+    }
+    return LoadPackagedLibrary(wlpName, 0);
+#else
     return LoadLibraryA(name);
+#endif
 }
 
 /*
diff --git a/xpath.c b/xpath.c
index 485d7747..c3501fb3 100644
--- a/xpath.c
+++ b/xpath.c
@@ -8481,7 +8481,7 @@ void
 xmlXPathLangFunction(xmlXPathParserContextPtr ctxt, int nargs) {
     xmlXPathObjectPtr val;
     xmlNodePtr cur;
-    xmlChar *theLang;
+    xmlChar *theLang = NULL;
     const xmlChar *lang;
     int ret = 0;
     int i;
